atef/widgets/config/page.py

```diff
@@ -0,0 +1,1664 @@
+        )
+        row_count = self.config_table.rowCount()
+        self.config_table.insertRow(row_count)
+        self.config_table.setRowHeight(row_count, config_row.sizeHint().height())
+        self.config_table.setCellWidget(row_count, 0, config_row)
+
+    def setup_row_buttons(
+        self,
+        config_row: ConfigurationGroupRowWidget,
+        config_item: AtefItem,
+    ):
+        self.setup_child_button(
+            button=config_row.child_button,
+            item=config_item,
+        )
+        self.setup_delete_row_button(
+            table=self.config_table,
+            item=config_item,
+            row=config_row,
+        )
+
+    def remove_table_data(self, data: Configuration):
+        self.data.configs.remove(data)
+
+    def resize_config_table(self):
+        """
+        Make sure that the whole row widget is visible.
+        """
+        self.config_table.setColumnWidth(0, self.config_table.width())
+
+    def showEvent(self, *args, **kwargs) -> None:
+        """
+        Override showEvent to update the table column width on first show.
+        """
+        self.resize_config_table()
+        return super().showEvent(*args, **kwargs)
+
+    def resizeEvent(self, *args, **kwargs) -> None:
+        """
+        Override resizeEvent to update the table column width when we resize.
+        """
+        self.resize_config_table()
+        return super().resizeEvent(*args, **kwargs)
+
+    def move_config_row(self, source: int, dest: int):
+        """
+        Move the row at index source to index dest.
+
+        Rearanges the table, the file, and the tree.
+        """
+        # Skip if into the same index
+        if source == dest:
+            return
+        config_data = self.data.configs.pop(source)
+        self.data.configs.insert(dest, config_data)
+        # Rearrange the tree
+        config_item = self.tree_item.takeChild(source)
+        self.tree_item.insertChild(dest, config_item)
+        # Rearrange the table: need a whole new widget or else segfault
+        self.config_table.removeRow(source)
+        self.config_table.insertRow(dest)
+        config_row = ConfigurationGroupRowWidget(data=config_data)
+        self.setup_row_buttons(
+            config_row=config_row,
+            config_item=config_item,
+        )
+        self.config_table.setRowHeight(dest, config_row.sizeHint().height())
+        self.config_table.setCellWidget(dest, 0, config_row)
+
+    def table_drop_event(self, event: QDropEvent):
+        """
+        Monkeypatch onto the table to allow us to drag/drop rows.
+
+        Shoutouts to stackoverflow
+        """
+        if event.source() is self.config_table:
+            selected_indices = self.config_table.selectedIndexes()
+            if not selected_indices:
+                return
+            selected_row = selected_indices[0].row()
+            dest_row = self.config_table.indexAt(event.pos()).row()
+            if dest_row == -1:
```

Comment:
I was testing the neat drag and drop functionality and found that for a simple one like `demo-configs/lfe.json`, dragging one of the configs to the end of the table caused it to disappear. 

Seems like the row count may be off-by-one, making it show up past the visible number of rows. For this single case, this worked:
```python
            if dest_row == -1:
                dest_row = self.config_table.rowCount() - 1
```
But hmm, it does make me wonder if `dest_row > source_row`, `dest_row` should be decremented by one (to account for the removed row) in general?

Reply:
> dragging one of the configs to the end of the table caused it to disappear

oh no I thought I tested this thoroughly

Reply:
I originally had some decrement logic in for dest_row > source_row but it seemed to break the table. When I removed it I thought I had everything working but I guess it's time to revisit.

Reply:
Hmm, strange...

Reply:
The drag and drop feels pretty robust with the tweak you suggested above. I don't fully understand why the decrement isn't needed for the dest > source case but it's definitely clear why placing the widget into a row outside the bounds of the table would break things- the last row is row count - 1, not row count. I guess when I was originally testing this I didn't have the table sized to be big or didn't drag into the blank space below the bottom row of the table, since all the drags within the bounds of the rows did work without subtracting 1 here.

Reply:
Sounds good enough to me! Thanks for double-checking the other scenario